sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1250,7 +1252,38 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
    */
   override def visitTableName(ctx: TableNameContext): LogicalPlan = withOrigin(ctx) {
     val tableId = visitMultipartIdentifier(ctx.multipartIdentifier)
-    val table = mayApplyAliasPlan(ctx.tableAlias, UnresolvedRelation(tableId))
+    val properties = new util.HashMap[String, String]
+    if (ctx.asOf != null && ctx.asOf.version != null) {
+      properties.put(TableCatalog.PROP_VERSION, ctx.asOf.version.getText)
```

Comment:
Should we also check `version` to be a non-negative `int`/`long`? (I guess `int` should be enough).

Reply:
`version` actually can't be negative. I defined `version=INTEGER_VALUE`.  `INTEGER_VALUE` is for positive number only.

Reply:
Yes I agree, but in parser, we don't have any restriction for the `version` to be a valid `int` right? The parser just restricts the version to be a string only contains numbers - `[0-9]`. It is a nit but it might lead to surprising behavior in system when query sets a `version` to be larger than maximal int. I agree this is rare, but we can just check `ctx.asOf.version.getText.toInt` here to make sure the invalid `version` won't pass to downstream code and can fail earlier.

```
fragment DIGIT
    : [0-9]
    ;

INTEGER_VALUE
    : DIGIT+
    ;
```

Reply:
Thanks for the suggestion. I am still not quite sure yet what this `VERSION` should be.  I think it is a snapshot id with type long. I will find out.

Reply:
Sorry for the late reply. I didn't have a chance earlier to find out what type the VERSION should be. This was just fixed.